types/src/traits/network.rs

```diff
@@ -109,105 +106,147 @@ pub enum NetworkError {
+    async fn shut_down_cc(&self) -> ();
+
+    /// broadcast message to those listening on the communication channel
+    /// non-blocking
+    async fn broadcast_message_cc(
         &self,
         message: Message<TYPES, LEAF, PROPOSAL>,
-    ) -> Result<(), NetworkError>;
+        election: &ELECTION,
+        view_number: TYPES::Time,
+    ) -> Result<RequestId, NetworkError>;
 
     /// Sends a direct message to a specific node
-    async fn message_node(
+    /// non-blocking
+    async fn direct_message_cc(
         &self,
         message: Message<TYPES, LEAF, PROPOSAL>,
         recipient: TYPES::SignatureKey,
-    ) -> Result<(), NetworkError>;
-
-    /// Provides a future for the next received message of `transmit_type`
-    ///
-    /// Will unwrap the underlying `NetworkMessage`
-    /// NOTE:
-    /// Deprecated because using this function on the centralized server implementation
-    /// of `NetworkingImplementation` will result in dropped messages. This is due to
-    /// an requirement of the implementation to block
-    /// until a single message is received or the channel is closed.
-    /// Use `recv_msgs` instead does not block and will return no messages.
-    #[deprecated]
-    async fn recv_msg(
-        &self,
-        transmit_type: TransmitType,
-    ) -> Result<Message<TYPES, LEAF, PROPOSAL>, NetworkError>;
+    ) -> Result<RequestId, NetworkError>;
 
     /// Moves out the entire queue of received messages of 'transmit_type`
     ///
     /// Will unwrap the underlying `NetworkMessage`
-    async fn recv_msgs(
+    /// blocking
+    async fn recv_msgs_cc(
         &self,
         transmit_type: TransmitType,
     ) -> Result<Vec<Message<TYPES, LEAF, PROPOSAL>>, NetworkError>;
 
-    /// Node's currently known to the networking implementation
-    ///
-    /// Kludge function to work around leader election
-    async fn known_nodes(&self) -> Vec<TYPES::SignatureKey>;
+    /// look up a node
+    /// non-blocking
+    async fn lookup_node_cc(&self, pk: TYPES::SignatureKey) -> Result<RequestId, NetworkError>;
+
+    /// cancel a message
+    /// if message completed or already cancelled, immediately returns
+    /// blocking
+    async fn cancel_msg_cc(&self, cancel_id: RequestId) -> Result<(), NetworkError>;
+
+    /// returns the status of a message
+    /// non-blocking
+    async fn msg_status_cc(&self, cancel_id: RequestId) -> RequestStatus;
+}
 
-    /// Returns a list of changes in the network that have been observed. Calling this function will clear the internal list.
-    async fn network_changes(
+/// common traits we would like olur network messages to implement
+pub trait NetworkMsg:
+    Serialize + for<'a> Deserialize<'a> + Clone + std::fmt::Debug + Sync + Send + 'static
+{
+}
+
+/// represents a networking implmentration
+/// exposes low level API for interacting with a network
+/// intended to be implemented for libp2p, the centralized server,
+/// and memory network
```

Comment:
I'm a bit confused about why we have the `ConnectedNetwork` trait specifically implemented for libp2p, the centralized server, and the memory network, but the implementation of each function is just called by the corresponding function of `CommunicationChannel`. Isn't this the same as putting the implementations directly in `CommunicationChannel` and removing `ConnectedNetwork`?

I might have missed something. I thought we wanted these two traits to support more than one instance of the networking implementation, but I'm not sure how this implementation fits into this idea.